### DISCUSSION TITLE
How shall we approach missing Ops in TorchSharp?

### DISCUSSION THREAD

**Discussion by GeorgeS2019** [K024/llm-sharp#3] [General]
#### TorchSharp
https://github.com/dotnet/TorchSharp/issues/790
https://github.com/dotnet/TorchSharp/issues/1081
https://github.com/dotnet/TorchSharp/pull/867

https://github.com/dotnet/TorchSharp/discussions/610

#### onnxruntime-extensions
https://github.com/microsoft/onnxruntime-extensions/issues/468

@K024 
I still have not developed an overview with Ops in TorchSharp

We really appreciate your effort to support TorchSharp community in this aspect. 

If you have idea the framework we need to work towards a community effort how to approach the missing Ops, please share and help us organize this topic better.

**Comment by K024** [comment]
There seems to be no magic. All you have to do is to export all the extra ops functions as a normal dll/so file and put that into `runtimes/<rid>/native/` in the project folder and load them with `[DllImport("xxx.dll")]`. I'm using python with `torch.utils.cpp_extensions` to setup the build pipeline. Code can be found in the [NativeOps](https://github.com/K024/llm-sharp/blob/main/NativeOps) folder. Pointers are used intensively and should be handled carefully.

**Comment by GeorgeS2019** [comment]
Need your view on these [Contributed Ops](https://github.com/microsoft/onnxruntime/tree/main/onnxruntime/contrib_ops) in Onnxruntime
- CPU
- CUDA

These are curated Ops contributed, AFAIK, from industry partners.

Under what circumstances would you think it make sense to bring these ops to Torchsharp?

Many of these Ops are contributed Deep NLP tokenizers. Perhaps some of which are related to your listed ones

Introduce more tokenizers

-  BPE
-  BPE (SentencePiece)
-  Tiktoken (BPE)
-  Unigram (SentencePiece)
-  WordPiece

I still do not have the complete overview yet,

I am not sure if these Contributed Ops are accessible through c# Onnxruntime API or they are limited to Python Onnxruntime API.

I appreciate your effort to bring more native ops to Torchsharp and show us the possible workflow how to do that.

**Reply by K024** [reply]
The tokenizers in this repo is purely implemented in C#, and have nothing to do with custom operators. If you'd like to introduce some C++ code into your project, compile it into DLL.
